pkg/k8s/watchers/endpoint.go

```diff
@@ -85,10 +88,80 @@ func (k *K8sWatcher) updateK8sEndpointV1(oldEP, newEP *slim_corev1.Endpoints, sw
 	if option.Config.BGPAnnounceLBIP {
 		k.bgpSpeakerManager.OnUpdateEndpoints(newEP)
 	}
+	k.addKubeAPIServerServiceEPs(newEP)
 	return nil
 }
 
 func (k *K8sWatcher) deleteK8sEndpointV1(ep *slim_corev1.Endpoints, swg *lock.StoppableWaitGroup) error {
 	k.K8sSvcCache.DeleteEndpoints(ep, swg)
 	return nil
 }
+
+func (k *K8sWatcher) handleKubeAPIServerServiceEPChanges(desiredIPs map[string]struct{}) {
+	// We must perform a diff on the ipcache.IdentityMetadata map in order to
+	// figure out which IPs are stale and should be removed, before we inject
+	// new IPs into the ipcache. The reason is because kube-apiserver will
+	// constantly reconcile this specific object, even when it's been deleted;
+	// effectively, this means we can avoid listening for the delete event.
+	// Therefore, any changes to this specific object can be handled in a
+	// "flattened" manner, since the most up-to-date form of it will be an add
+	// or update event. The former is sent when Cilium is syncing with K8s and
+	// the latter is sent anytime after.
+	//
+	// For example:
+	//   * if a backend is removed or updated, then this will be in the form of
+	//     an update event.
+	//   * if the entire object is deleted, then it will quickly be recreated
+	//     and this will be in the form of an add event.
+	ips := ipcache.FilterMetadataByLabels(labels.LabelKubeAPIServer)
+	currentIPs := make(map[string]struct{}, len(ips))
```

Comment:
I used `oldAPIServerIPs` instead to be a bit more explicit; this is the 'current' / existing list from the ipcache, which will be compared against the incoming 'desired' set of IPs to associate with the kube-apiserver, and then we'll generate the diffs so that we can remove IPs that are no longer associated with the apiserver and add IPs that should now be associated with the apiserver.